src/createSchemaForm.tsx

```diff
@@ -141,6 +141,143 @@ function propsMapToObect(propsMap: PropsMapping) {
   return r;
 }
 
+type SchemaShape<SchemaType extends RTFSupportedZodTypes | ZodEffects<any, any>> = ReturnType<UnwrapEffects<SchemaType>['_def']["shape"]>
+
+type IndexOfSchemaInMapping<
+ Mapping extends FormComponentMapping, 
+ SchemaType extends RTFSupportedZodTypes | ZodEffects<any, any>,
+ key extends keyof z.infer<UnwrapEffects<SchemaType>>
+> = IndexOf<UnwrapMapping<Mapping>, readonly [
+  IndexOfUnwrapZodType<SchemaShape<SchemaType>[key]>,
+  any
+]>;
+
+type GetTupleFromMapping<
+Mapping extends FormComponentMapping, 
+SchemaType extends RTFSupportedZodTypes | ZodEffects<any, any>, 
+key extends keyof z.infer<UnwrapEffects<SchemaType>>
+> = 
+IndexOfSchemaInMapping<Mapping, SchemaType, key> extends never ? never : Mapping[IndexOfSchemaInMapping<Mapping, SchemaType, key>];
+
+type Prev = [never, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
+export type PropType<
+  Mapping extends FormComponentMapping,
+  SchemaType extends RTFSupportedZodTypes | ZodEffects<any, any>,
+  PropsMapType extends PropsMapping = typeof defaultPropsMap,
+  // this controls the depth we allow TS to go into the schema. 2 is enough for most cases, but we could consider exposing this as a generic to allow users to control the depth
+  Level extends Prev[number] = 2 
+> = [Level] extends [never] ? never : RequireKeysWithRequiredChildren<
+ Partial<{
+    [key in keyof z.infer<UnwrapEffects<SchemaType>>]:
+    GetTupleFromMapping<Mapping, SchemaType, key> extends never 
+    ? UnwrapEffects<SchemaType>["shape"][key] extends z.AnyZodObject 
+      ?  PropType<Mapping, UnwrapEffects<SchemaType>["shape"][key], PropsMapType, Prev[Level]> 
+      : UnwrapEffects<SchemaType>["shape"][key] extends z.ZodArray<any>  
+        ? PropType<Mapping, UnwrapEffects<SchemaType>["shape"][key]['element'], PropsMapType, Prev[Level]>  
+        : never 
+    : GetTupleFromMapping<Mapping, SchemaType, key> extends readonly [any, any] // I guess this tells typescript it has a second element? errors without this check.
+    ? Omit<
+        ComponentProps<
+        GetTupleFromMapping<Mapping, SchemaType, key>[1]
+        >,
+        PropsMapType[number][1]
+      > & ExtraProps
+    :  never
+  }>
+>
+
+type RenderedFieldMap<SchemaType extends AnyZodObject | ZodEffects<any, any>> = {
+  [key in keyof z.infer<UnwrapEffects<SchemaType>>]: UnwrapEffects<SchemaType>['shape'][key] extends z.AnyZodObject?
+   RenderedFieldMap<UnwrapEffects<SchemaType>['shape'][key]> : 
+     UnwrapEffects<SchemaType>['shape'][key] extends z.ZodArray<any> ?
+     UnwrapEffects<SchemaType>['shape'][key]['element'] extends z.AnyZodObject ? 
+      RenderedFieldMap<UnwrapEffects<SchemaType>['shape'][key]['element']>[] :
+     JSX.Element[] :
+      JSX.Element;
+};
+
+
+type RTFFormProps<Mapping extends FormComponentMapping, SchemaType extends z.AnyZodObject | ZodEffects<any, any>, PropsMapType extends PropsMapping = typeof defaultPropsMap,FormType extends FormComponent = "form"> = {
```

Comment:
had to pull this out to make the return type "explicit" for TS also it's a bit nicer maybe since it's so huge

Reply:
nice this is actually a lot cleaner